src/librustc_mir/interpret/place.rs

```diff
@@ -1040,21 +1040,18 @@ where
     ) -> InterpResult<'tcx> {
         let variant_scalar = Scalar::from_u32(variant_index.as_u32()).into();
 
+        if dest.layout.for_variant(self, variant_index).abi.is_uninhabited() {
+            throw_ub!(InvalidDiscriminant(variant_scalar));
+        }
+
         match dest.layout.variants {
-            layout::Variants::Single { index } => {
-                if index != variant_index {
```

Comment:
Please keep that check (as an `assert!` is possible)

Reply:
I may have suggested this too, as codegen does have an `assert_eq!(index, variant_index)` here IIRC.

Reply:
Ah, yes. You're correct

https://github.com/rust-lang/rust/blob/fe969f4ec67a5bd317818cbc7a648c5795af15fa/src/librustc_codegen_ssa/mir/place.rs#L339-L341